### PR TITLE
Set-Metadata HTTP filter

### PR DESCRIPTION
**Pull request by aguinet**
Commit Message: HTTP filter that can set or update dynamic metadata.

Additional Description:

This extension simply allows to add user-defined dynamic metadata. This is intended to be used with matchers, so that the "result" of a matching process can imply specific metadata.

See issue #16266 

Risk Level: Medium
Testing: Unit tests for the extension are added
Docs Changes: A small documentation for the extension is added in `docs/root/configuration/http/http_filters`
Release Notes: N/A
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/BUILD
Comment (htuch) [inline api/BUILD:212]: Need to update https://github.com/envoyproxy/envoy/blob/main/CODEOWNERS

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [inline source/extensions/filters/http/set_metadata/set_metadata_filter.cc:41]: I'm not sure exactly how `MergeFrom` and standard `Struct` merging works. In https://github.com/envoyproxy/envoy/issues/16266#issuecomment-832929796 it's claimed it does the wrong thing and so we need a custom merger here. Have you validated this empirically?

I think this merger should be a library if needed and placed in `source/common/protobuf`, since we probably want this as a general `StructUtils` kind of thing. But I can't tell from reading the description of `MergeFrom` if it doesn't already do what you need.

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [inline source/extensions/filters/http/set_metadata/set_metadata_filter.cc:64]: Nit: `org_fields` (and other places with camel/snake case confusion in local vars.

File: api/envoy/extensions/filters/http/set_metadata/v3/set_metadata.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/set_metadata/v3/set_metadata.proto:27]: RST link to docs providing details of the merge semantics.

File: api/envoy/extensions/filters/http/set_metadata/v3/set_metadata.proto
Comment (htuch) [inline api/envoy/extensions/filters/http/set_metadata/v3/set_metadata.proto:7]: Probably don't need `migrate.proto`

**Review by htuch — COMMENTED** [review]
Looks good generally, just want to verify the details around merge.

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (aguinet) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc:41]: I tried `MergeFrom`, and indeed validated empirically that it does not do what we want *for structures*, as per the example you mentionned. For lists, it is just appending the values (which is what we want), as we can see here https://github.com/protocolbuffers/protobuf/blob/465acaab6651a158592b06605a38fed373512f23/src/google/protobuf/struct.pb.cc#L1023 and here https://github.com/protocolbuffers/protobuf/blob/465acaab6651a158592b06605a38fed373512f23/src/google/protobuf/repeated_field.h#L1826-L1867 . I also validated this empirically and indirectly in the unit tests. I also couldn't find any proper documentation on this...

I'm okay to move this into `source/common/protobuf`, with a separate commit and unit tests!

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (aguinet) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc]: Should be fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/9d26ca29521869cfb630b3e7f02d271b7ff142bc

File: api/envoy/extensions/filters/http/set_metadata/v3/set_metadata.proto
Comment (aguinet) [reply api/envoy/extensions/filters/http/set_metadata/v3/set_metadata.proto]: Should be fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/9d26ca29521869cfb630b3e7f02d271b7ff142bc

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc]: OK, let's do that then. This kind of protobuf merge is probably of interest to @envoyproxy/api-shepherds; do we want to be using this as a standard merge feature in the APIs when combining Structs?

Comment (aguinet) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc]: So commit https://github.com/envoyproxy/envoy/pull/16400/commits/acb5e25389b4879cbc328dd85464689f6be9b616 introduce `StructUtil::update`, with the related tests. And commit https://github.com/envoyproxy/envoy/pull/16400/commits/7540c2a07e1a8d4f9ac70006d6ffda037c051e87 uses it for the set-metadata filter.

Writing the tests for `StructUtil::update` made me realize that I didn't take into account the case where two keys exist in the original and new object but do not have the same type. I took the strategy of "last-one-win" here (that is taking the value of the new object). I modified the documentation of set-metadata to also discuss this case.

I also just realized I forgot the doc of the `StructUtil::update` API itself, I'm writing a commit to add it.

BTW, when I'll rebase/squash I'll obviously put the commit about `StructUtil::update` first in history!

Comment (aguinet) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc]: API documentation is in https://github.com/envoyproxy/envoy/pull/16400/commits/4e846f0d5fe4dc9551ea951947d0337841c2b9d6

File: source/common/protobuf/utility.cc
Comment (htuch) [inline source/common/protobuf/utility.cc:1047]: Nit: `const ProtobufWkt::Struct& with`

Comment (aguinet) [reply source/common/protobuf/utility.cc]: Sorry for this... Fixed in 56ea00fc5efb3a221d8dc370bdf5e78263216900

**Comment by aguinet** [comment]
@htuch it looks like this conflicts with main. Should I be rebase on main and resquash into two commits? (one for `StructUtil::update` and one for the extensions)

**Comment by htuch** [comment]
@aguinet in general, you should have a flow that you never rebase after creating a PR. Instead do local merges from `origin/main`.

File: api/BUILD
Comment (htuch) [reply api/BUILD:212]: Yes, Snow and you.

**Review by htuch — COMMENTED** [review]
Thanks. The merging now makes more sense. One question I have is how this extension relates to the `header_to_metadata` filter? https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto

It seems that really we want both to behave similarly, and if we were designing from scratch they would be somewhat unified. We also have other filters, e.g. ext_authz, that set dynamic metadata base on return values from remote servers. Ideally these have aligned merge semantics.

We don't need to boil the ocean here, but having some explanation and game plan to get consistent behaviors would be helpful.

**Comment by aguinet** [comment]
> Thanks. The merging now makes more sense. One question I have is how this extension relates to the `header_to_metadata` filter? https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/http/header_to_metadata/v3/header_to_metadata.proto
> 
> It seems that really we want both to behave similarly, and if we were designing from scratch they would be somewhat unified. We also have other filters, e.g. ext_authz, that set dynamic metadata base on return values from remote servers. Ideally these have aligned merge semantics.
> 
> We don't need to boil the ocean here, but having some explanation and game plan to get consistent behaviors would be helpful.

These are good questions I tried to resolve before starting this. Let me first explain my use case. What we'd like to do is leverage the matching engine to add "tags", and log this tag and/or take decisions according to these tags. This is configurable by the user in the end, and would be transformed to an envoy configuration file.

As an example, I could add the tag "tag0" if the cookie "mycookie" matches a given regular expression, and another tag "tag1" if the IP comes from AS3215. I would like these two tags to be logged, and let's say block the request if the two of them are there.

In order to reuse the matching engine, we would configure HTTP filters like this:

```
# Wrapped into a first matcher
- filter0: 
    @type: set_metadata
    namespace: mynamespace
    value:
      tags:
       tag0: 1
#  Wrapped into a second matcher
- filter1:
    @type: set_metadata 
    namespace: mynamespace
    value:
      tags:
        tag1: 1
```

In order to get the two tags, I need the current merging strategy.

Now about `header_to_metadata`. From my understanding, it's just setting strings (potentially base64 ones) or integers values, taken from headers. There are no issue of potentially merging two different recursive structures. But I might be missing something here.

I just checked `ext_authz`, the only place I found it setting metadata is here: https://github.com/envoyproxy/envoy/blob/1f4ca4c0d395360845d0453a949f6c53acb147cb/source/extensions/filters/http/ext_authz/ext_authz.cc#L211-L214 . This looks like there are no merging strategy involved, but I don't know the use cases enough to know if that even makes sens!

**Comment by aguinet** [comment]
> @aguinet in general, you should have a flow that you never rebase after creating a PR. Instead do local merges from `origin/main`.

Okay. So I'll end up with a merge commit, right? Should I squash all of this before merging?

**Comment by htuch** [comment]
Thanks for the explanation. I agree this makes sense. @envoyproxy/api-shepherds I'd like us to think about adopting this as the standard merge strategy going forward, so we don't have inconsistent merging when filters are performing dynamic metadata update.

> Okay. So I'll end up with a merge commit, right? Should I squash all of this before merging?

Yes, add a merge commit. You can't squash (since this is a rebase) once a PR has started. We will squash everything prior to the main merge once approved.

File: docs/root/configuration/http/http_filters/set_metadata_filter.rst
Comment (htuch) [inline docs/root/configuration/http/http_filters/set_metadata_filter.rst:10]: @envoyproxy/api-shepherds this differs from `MergeFrom` in that it supports merging structs together, e.g. if you have a disjoint namespace things compose as expected. I think this makes sense, but flagging this for consideration in case there is some alternate merging preferences.

File: source/common/protobuf/utility.cc
Comment (htuch) [inline source/common/protobuf/utility.cc:1050]: Again here with the `auto const&`; I think you want `const auto&`

File: source/common/protobuf/utility.cc
Comment (htuch) [inline source/common/protobuf/utility.cc:1054]: Nit: `const auto&`

File: source/common/protobuf/utility.cc
Comment (htuch) [inline source/common/protobuf/utility.cc:1076]: Nit: `const auto& vals`

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [inline source/extensions/filters/http/set_metadata/set_metadata_filter.cc:28]: `const std::string& metadata_namespace`

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [inline source/extensions/filters/http/set_metadata/set_metadata_filter.cc:31]: `const ProtobufWkt::Struct& to_merge`

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (htuch) [inline test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc:48]: Nit: const these locals.

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (htuch) [inline test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc:47]: `checkKeyInt`

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (htuch) [inline test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc:74]: Nit: `const auto&` all these locals here and below

File: source/extensions/filters/http/set_metadata/config.cc
Comment (htuch) [inline source/extensions/filters/http/set_metadata/config.cc:18]: Is there a test covering this?

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [inline source/extensions/filters/http/set_metadata/set_metadata_filter.cc:39]: Nit: don't think this is covered in tests.

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (htuch) [inline test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc:58]: Do the headers have any value in these tests? I'm not seeing anything making use of them.

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks; a bunch of API and code comments to address, I think it's in the right shape.
/wait

File: source/common/protobuf/utility.cc
Comment (aguinet) [reply source/common/protobuf/utility.cc:1054]: kind_case() returns a scalar (https://github.com/protocolbuffers/protobuf/blob/7e95c64dfbd9f725f005c83470d44140f58ce8d0/src/google/protobuf/struct.pb.h#L497 & https://github.com/protocolbuffers/protobuf/blob/7e95c64dfbd9f725f005c83470d44140f58ce8d0/src/google/protobuf/struct.pb.h#L317), we can just copy its value no?

File: source/common/protobuf/utility.cc
Comment (aguinet) [reply source/common/protobuf/utility.cc]: Fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c

File: source/common/protobuf/utility.cc
Comment (aguinet) [reply source/common/protobuf/utility.cc]: Fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (aguinet) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc]: Fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c (used `const absl::string_view` as this is what ̀`metadataNamespace` returns).

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (aguinet) [reply test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc]: Fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (aguinet) [reply test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc]: Fixed in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c (and in other tests)

File: source/extensions/filters/http/set_metadata/config.cc
Comment (aguinet) [reply source/extensions/filters/http/set_metadata/config.cc:18]: Added in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c (config_test.cc)

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (aguinet) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc:39]: Added in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c , by calling `decodeData` in `runFilter`

File: test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc
Comment (aguinet) [reply test/extensions/filters/http/set_metadata/set_metadata_filter_test.cc]: Indeed they are useless. Reduced and factorized in https://github.com/envoyproxy/envoy/pull/16400/commits/cfad0846aa208507d3e5450807bda6512209028c

File: api/BUILD
Comment (aguinet) [reply api/BUILD:212]: Sorry missed this answer. Done in d139579d9ed5c2a00090f4e75f61df26d0d8eda9 !

File: source/extensions/filters/http/set_metadata/set_metadata_filter.cc
Comment (htuch) [reply source/extensions/filters/http/set_metadata/set_metadata_filter.cc:31]: Ping on this

File: test/common/protobuf/utility_test.cc
Comment (htuch) [inline test/common/protobuf/utility_test.cc:2023]: More `const` issues. Sorry if this seems nit-picky, but this style of `const` is different to the Envoy code base and consistency is important. There are some rare exceptions and we should fix them.

Can you just grep for `const&` in your difference from `main` and fix them all? Thanks.

**Comment by htuch** [comment]
LGTM modulo nits and a main merge for conflict.

File: test/common/protobuf/utility_test.cc
Comment (aguinet) [reply test/common/protobuf/utility_test.cc:2023]: I'm sorry I thought I had checked them all... I completely understand the style consistency. Will indeed recheck this properly!

Comment (aguinet) [reply test/common/protobuf/utility_test.cc]: Fixed in 45ca93383ba5ad5291a6fb9bae97b4d58ebb1fb5 .

```
$ git diff main...feature/ext_add_metadata |grep 'const&' |wc -l
0
```

Hopefully I should be fine!

**Review by htuch — DISMISSED** [review]
LGTM, thanks!

Needs main merge.

**Comment by aguinet** [comment]
@htuch main merged into the branch! Is there something else I need to do?

Thanks for the time you took for all the reviews and remarks!

**Comment by htuch** [comment]
@phlax do you know what is going on with this glint failure?

**Comment by phlax** [comment]
https://dev.azure.com/cncf/envoy/_build/results?buildId=76293&view=logs&j=c5dd2866-6ab3-5f3c-3a44-4cef0ec909b5&t=a9eb66d6-8944-5769-b3f7-476949dadcb8&l=60

**Comment by aguinet** [comment]
> https://dev.azure.com/cncf/envoy/_build/results?buildId=76293&view=logs&j=c5dd2866-6ab3-5f3c-3a44-4cef0ec909b5&t=a9eb66d6-8944-5769-b3f7-476949dadcb8&l=60

Should be fixed with 0469f862691473bdc1e0630ea4e043f16b7b5871 . Added a configuration to my vim to highlight these...

**Comment by phlax** [comment]
> Added a configuration to my vim to highlight these...

i have my editor (emacs) automatically remove/fix on save

**Review by htuch — DISMISSED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@aguinet the clang-tidy failures seem legit, see https://dev.azure.com/cncf/envoy/_build/results?buildId=76440&view=logs&j=430ab721-bbba-5b3a-3c7b-ff111cc657fe&t=b6238472-0f30-5195-100c-623a3015b9ef

**Comment by aguinet** [comment]
@htuch yep, I've just pushed https://github.com/envoyproxy/envoy/pull/16400/commits/d82b20ab3239f5118b45a3aec2fc94fa5f49550c to fix it!

**Review by htuch — APPROVED** [review]

**Comment by aguinet** [comment]
\o/ Thanks a lot for the merge @htuch !

**Pull request merged**
